src/ringbuffer.h

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_RINGBUFFER_H
+#define BITCOIN_RINGBUFFER_H
+
+#include <array>
+#include <atomic>
+#include <cassert>
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+/**
+ * A threadsafe circular buffer.
+ *
+ * Allows optional blocking behavior instead of overwriting when at capacity.
+ */
+template<typename ET, int S, typename ST=int>
```

Comment:
Do we really need the size type template parameter? Especially if you are passing an int through as a value. `size_t` makes more sense to me since std::array is size_t and signed sizes (int) also make no sense.

Reply:
Yep, agree. Fixed.